File: flow/script/lib/transcribe.ts
/**
 * Audio transcription via GCP Vertex AI and Gemini fallback.
 *
 * Architecture:
 *   1. Try GCP Vertex AI first (project credits, recommended)
 *   2. Fallback to Gemini standard API if Vertex fails
 *
 * Implementation uses direct REST API calls to avoid google-genai
 * client initialization issues with API key + project combos.
 */

import { loadConfig, validateConfig, type TranscribeConfig } from "../config/transcribe.ts";
import { BINARIES } from "../config/paths.ts";

const PYTHON = BINARIES.ytDlp.replace("yt-dlp", "python3");

export interface TranscribeResult {
  text: string;
  backend: string;
  model: string;
}

/** Run Python script via subprocess */
async function runPython(script: string): Promise<string> {
  const cmd = new Deno.Command(PYTHON, {
    args: ["-c", script],
    stdout: "piped",
    stderr: "piped",
  });
  const { code, stdout, stderr } = await cmd.output();
  const out = new TextDecoder().decode(stdout);
  const err = new TextDecoder().decode(stderr);

  if (code !== 0) {
    throw new Error(`Python exited ${code}: ${err.slice(0, 500)}`);
  }
  return out.trim();
}

// ── GCS upload helper ─────────────────────────────────────────────────

/** Upload file to GCS bucket */
async function uploadToGcs(
  localPath: string,
  cfg: TranscribeConfig,
): Promise<string> {
  const script = `
import sys
import os
from google.cloud import storage

client = storage.Client(project=${JSON.stringify(cfg.gcpProjectId)})
bucket = client.bucket(${JSON.stringify(cfg.gcsBucket)})
blob_name = ${JSON.stringify(cfg.gcsPrefix)} + os.path.basename(${JSON.stringify(localPath)})
blob = bucket.blob(blob_name)
blob.upload_from_filename(${JSON.stringify(localPath)})
print(f"gs://${cfg.gcsBucket}/{blob_name}")
`;
  return await runPython(script);
}

// ── Vertex AI transcription via REST ──────────────────────────────────

/** Transcribe via Vertex AI REST API with inline audio */
async function transcribeVertex(
  audioPath: string,
  cfg: TranscribeConfig,
): Promise<string> {
  const script = `
import urllib.request
import json
import base64

api_key = ${JSON.stringify(cfg.gcpApiKey)}
audio_path = ${JSON.stringify(audioPath)}

with open(audio_path, "rb") as f:
    audio_b64 = base64.b64encode(f.read()).decode()

url = (
    f"https://${cfg.gcpLocation}-aiplatform.googleapis.com/v1/"
    f"projects/${cfg.gcpProjectId}/locations/${cfg.gcpLocation}/"
    f"publishers/google/models/${cfg.gcpModel}:generateContent"
    f"?key={api_key}"
)

data = {
    "contents": [{
        "role": "user",
        "parts": [
            {"text": "Transcribe this audio accurately."},
            {"inline_data": {"mime_type": "audio/mpeg", "data": audio_b64}}
        ]
    }],
    "generationConfig": {"temperature": 0, "maxOutputTokens": 8192}
}

req = urllib.request.Request(
    url,
    data=json.dumps(data).encode(),
    headers={"Content-Type": "application/json"}
)
resp = urllib.request.urlopen(req)
result = json.loads(resp.read())
text = result["candidates"][0]["content"]["parts"][0]["text"]
print(text)
`;
  return await runPython(script);
}

// ── Gemini standard API transcription ─────────────────────────────────

/** Transcribe via Gemini standard API with inline audio */
async function transcribeGemini(
  audioPath: string,
  cfg: TranscribeConfig,
): Promise<string> {
  const script = `
import urllib.request
import json
import base64

api_key = ${JSON.stringify(cfg.googleApiKey)}
audio_path = ${JSON.stringify(audioPath)}

with open(audio_path, "rb") as f:
    audio_b64 = base64.b64encode(f.read()).decode()

url = (
    "https://generativelanguage.googleapis.com/v1beta/"
    f"models/${cfg.geminiStdModel}:generateContent?key={api_key}"
)

data = {
    "contents": [{
        "parts": [
            {"text": "Transcribe this audio accurately."},
            {"inline_data": {"mime_type": "audio/mpeg", "data": audio_b64}}
        ]
    }],
    "generationConfig": {"temperature": 0, "maxOutputTokens": 8192}
}

req = urllib.request.Request(
    url,
    data=json.dumps(data).encode(),
    headers={"Content-Type": "application/json"}
)
resp = urllib.request.urlopen(req)
result = json.loads(resp.read())
text = result["candidates"][0]["content"]["parts"][0]["text"]
print(text)
`;
  return await runPython(script);
}

// ── Public API ────────────────────────────────────────────────────────

/**
 * Transcribe audio file.
 * Fault-tolerant: tries Vertex AI first, falls back to Gemini.
 */
export async function transcribeAudio(
  audioPath: string,
  preferredBackend?: "gcp-vertex" | "gemini",
): Promise<TranscribeResult> {
  const cfg = loadConfig();
  validateConfig(cfg);

  // Determine backend order
  const backends: Array<"gcp-vertex" | "gemini"> = [];
  if (preferredBackend) {
    backends.push(preferredBackend);
  } else if (cfg.transcriber) {
    backends.push(cfg.transcriber);
  } else {
    // Auto failover chain: vertex first, then gemini
    if (cfg.gcpApiKey) backends.push("gcp-vertex");
    if (cfg.googleApiKey) backends.push("gemini");
  }

  if (backends.length === 0) {
    throw new Error("No transcription backend available");
  }

  const errors: string[] = [];

  for (const backend of backends) {
    try {
      if (backend === "gcp-vertex") {
        console.log(`[INFO] Trying GCP Vertex AI (${cfg.gcpModel})...`);
        const text = await transcribeVertex(audioPath, cfg);
        return { text, backend, model: cfg.gcpModel };
      } else {
        console.log(`[INFO] Trying Gemini API (${cfg.geminiStdModel})...`);
        const text = await transcribeGemini(audioPath, cfg);
        return { text, backend, model: cfg.geminiStdModel };
      }
    } catch (e: any) {
      console.warn(`[WARN] ${backend} failed: ${e.message.slice(0, 200)}`);
      errors.push(`${backend}: ${e.message}`);
    }
  }

  throw new Error(
    `All transcription backends failed:\n${errors.join("\n")}`,
  );
}

File: flow/script_py/lib/transcribe.py
"""
Audio transcription via GCP Vertex AI and Gemini fallback.

Architecture:
  1. Try GCP Vertex AI first (project credits, recommended)
  2. Fallback to Gemini standard API if Vertex fails

Uses direct REST API calls to avoid google-genai client issues.
"""

import base64
import json
import urllib.request
from typing import Optional

from config.paths import BINARIES
from config.transcribe import TranscribeConfig, load_config

PROMPT = (
    "请仔细转写这段音频的内容。保持原文语言，添加标点，"
    "按语义分段，去除语气词和广告词，输出纯文本。"
)

PYTHON = BINARIES["python"]


def _build_vertex_script(audio_path: str, cfg: TranscribeConfig) -> str:
    return f'''
import urllib.request
import json
import base64

api_key = {json.dumps(cfg.gcp_api_key)}
audio_path = {json.dumps(audio_path)}

with open(audio_path, "rb") as f:
    audio_b64 = base64.b64encode(f.read()).decode()

url = (
    f"https://{cfg.gcp_location}-aiplatform.googleapis.com/v1/"
    f"projects/{cfg.gcp_project_id}/locations/{cfg.gcp_location}/"
    f"publishers/google/models/{cfg.gcp_model}:generateContent"
    f"?key={{api_key}}"
)

data = {{
    "contents": [{{
        "role": "user",
        "parts": [
            {{"text": "请仔细转写这段音频的内容。保持原文语言，添加标点，按语义分段，去除语气词和广告词，输出纯文本。"}},
            {{"inline_data": {{"mime_type": "audio/mpeg", "data": audio_b64}}}}
        ]
    }}],
    "generationConfig": {{"temperature": 0, "maxOutputTokens": 8192}}
}}

req = urllib.request.Request(
    url,
    data=json.dumps(data).encode(),
    headers={{"Content-Type": "application/json"}}
)
resp = urllib.request.urlopen(req)
result = json.loads(resp.read())
text = result["candidates"][0]["content"]["parts"][0]["text"]
print(text)
'''


def _build_gemini_script(audio_path: str, cfg: TranscribeConfig) -> str:
    return f'''
import urllib.request
import json
import base64

api_key = {json.dumps(cfg.google_api_key)}
audio_path = {json.dumps(audio_path)}

with open(audio_path, "rb") as f:
    audio_b64 = base64.b64encode(f.read()).decode()

url = (
    "https://generativelanguage.googleapis.com/v1beta/"
    f"models/{cfg.gemini_std_model}:generateContent?key={{api_key}}"
)

data = {{
    "contents": [{{
        "parts": [
            {{"text": "请仔细转写这段音频的内容。保持原文语言，添加标点，按语义分段，去除语气词和广告词，输出纯文本。"}},
            {{"inline_data": {{"mime_type": "audio/mpeg", "data": audio_b64}}}}
        ]
    }}],
    "generationConfig": {{"temperature": 0, "maxOutputTokens": 8192}}
}}

req = urllib.request.Request(
    url,
    data=json.dumps(data).encode(),
    headers={{"Content-Type": "application/json"}}
)
resp = urllib.request.urlopen(req)
result = json.loads(resp.read())
text = result["candidates"][0]["content"]["parts"][0]["text"]
print(text)
'''


def _run_python(script: str) -> str:
    """Run Python script via subprocess."""
    import subprocess
    result = subprocess.run(
        [PYTHON, "-c", script],
        capture_output=True,
        text=True,
        timeout=300,
    )
    if result.returncode != 0:
        raise RuntimeError(f"Python exited {result.returncode}: {result.stderr[:500]}")
    return result.stdout.strip()


def transcribe_audio(
    audio_path: str,
    preferred_backend: Optional[str] = None,
) -> dict[str, str]:
    """
    Transcribe audio file.
    Fault-tolerant: tries Vertex AI first, falls back to Gemini.
    """
    cfg = load_config()
    cfg.validate()

    # Determine backend order
    backends: list[str] = []
    if preferred_backend:
        backends.append(preferred_backend)
    elif cfg.transcriber:
        backends.append(cfg.transcriber)
    else:
        if cfg.gcp_api_key:
            backends.append("gcp-vertex")
        if cfg.google_api_key:
            backends.append("gemini")

    if not backends:
        raise RuntimeError("No transcription backend available")

    errors: list[str] = []

    for backend in backends:
        try:
            if backend == "gcp-vertex":
                print(f"[INFO] Trying GCP Vertex AI ({cfg.gcp_model})...")
                script = _build_vertex_script(audio_path, cfg)
                text = _run_python(script)
                return {"text": text, "backend": backend, "model": cfg.gcp_model}
            else:
                print(f"[INFO] Trying Gemini API ({cfg.gemini_std_model})...")
                script = _build_gemini_script(audio_path, cfg)
                text = _run_python(script)
                return {"text": text, "backend": backend, "model": cfg.gemini_std_model}
        except Exception as e:
            msg = f"{backend} failed: {str(e)[:200]}"
            print(f"[WARN] {msg}")
            errors.append(msg)

    raise RuntimeError(f"All transcription backends failed:\n" + "\n".join(errors))
